sbol_utilities/default_components.py

```diff
@@ -0,0 +1,148 @@
+    terminator_seq.encoding = 'https://identifiers.org/edam:format_1207'
+
+    terminator = sbol3.Component(name, sbol3.SBO_DNA)
+    terminator.name = name
+    terminator.description = description
+    terminator.roles. append(sbol3.SO_TERMINATOR)
+    terminator.sequences.append(terminator_seq)
+    return tuple([terminator, terminator_seq])
+
+def deg_tag(name: str, description: str, sequence: str)-> Tuple[sbol3.Component, sbol3.Sequence]:
+    """
+    Creates a degradation tag and a sequence Components with a given name, description, and sequence.
+    """
+    deg_tag_seq = sbol3.Sequence(f'{name}_seq')
+    deg_tag_seq.elements= sequence
+    deg_tag_seq.encoding = 'https://identifiers.org/edam:format_1207'
+
+    deg_tag = sbol3.Component(name, sbol3.SBO_DNA)
+    deg_tag.name = name
+    deg_tag.description = description
+    deg_tag.roles. append(tyto.SO.protein_stability_element)
+    deg_tag.sequences.append(deg_tag_seq)
+    return tuple([deg_tag, deg_tag_seq])
+
+def plasmid_vector(name: str, description: str, sequence: str)-> Tuple[sbol3.Component, sbol3.Sequence]:
+    """
+    Creates a plasmid vector or backbone and a sequence Components with a given name, description, and sequence.
+    """
+    plasmid_vector_seq = sbol3.Sequence(f'{name}_seq')
+    plasmid_vector_seq.elements= sequence
+    plasmid_vector_seq.encoding = 'https://identifiers.org/edam:format_1207'
+
+    plasmid_vector = sbol3.Component(name, sbol3.SBO_DNA)
+    plasmid_vector.name = name
+    plasmid_vector.description = description
+    plasmid_vector.roles. append(tyto.SO.plasmid_vector)
+    plasmid_vector.sequences.append(plasmid_vector_seq)
+    return tuple([plasmid_vector, plasmid_vector_seq])
+
+def engineered_region(name: str, description: str, sub_components: List[sbol3.SubComponent])-> sbol3.Component:
+    """
+    Creates an engineered region Componentswith a given name, description, and sub components.
+    """
+    #sequence = blunt or assembly?
+    #engineered_region_seq = sbol3.Sequence(f'{name}_seq')
+    #engineered_region_seq.elements= sequence
+    #engineered_region_seq.encoding = 'https://identifiers.org/edam:format_1207'
+
+    engineered_region = sbol3.Component(name, sbol3.SBO_DNA)
+    engineered_region.name = name
+    engineered_region.description = description
+    engineered_region.roles.append(sbol3.SO_ENGINEERED_REGION)
+    #create subcomponents if Components are passed  
+    for x in sub_components:
+        engineered_region.features.append(x)
+    #fix order
+    if len(engineered_region.features) > 1:
+            for i in range(len(engineered_region.features)-1):
+                engineered_region.constraints = [sbol3.Constraint(sbol3.SBOL_PRECEDES, engineered_region.features[i], engineered_region.features[i+1])]
+    else: pass
+    #engineered_region.sequences.append(engineered_region_seq)
+    return engineered_region # , engineered_region_seq
+
+def media(name: str, description: str)-> sbol3.Component:
```

Comment:
How about having the recipe for the media optionally passed as a `Dict(Component,Measure)`?

Reply:
I like the idea of having the recipe of the media. How can we store that info into SBOL3? My first idea is to create Components  of the recipe items and append the Measure to that Component measures , then make SubComponents from the recipe items and append them as features on the media Component. What do you think?

Reply:
Section 7.9.1 of the specification already provides this as a recommended practice.

I would suggest that in some cases it would make sense to use SubComponent and in others to use ExternalReference. For some URIs, it would be obvious which to use (e.g., ChEBI is ExternalReference, a Component in the document is SubComponent). For other URIs, it would not be obvious, however, and might need to be provided by the user in some manner, e.g., a wrapping function.

Reply:
I have a proposal implementation now. The recipe would be a dictionary where the key is a Component and the value is a list of a number and a measure unit URI like this:
```
peptone = sbol3.Component('Bacto_Peptone', 
            tyto.SBO.functional_entity,
            name = 'Bacto_Peptone',
            derived_from = ['https://www.thermofisher.com/order/catalog/product/211820'])

nacl = sbol3.Component('NaCl', 
            tyto.SBO.functional_entity,
            name = 'NaCl',
            derived_from = ['https://www.sigmaaldrich.com/AU/en/product/sigald/s9888'])

yeast_extract = sbol3.Component('Yeast_Extract', 
            tyto.SBO.functional_entity,
            name = 'Yeast_Extract',
            derived_from = ['https://www.thermofisher.com/order/catalog/product/212720'])

recipe = {
    peptone:[10,tyto.OM.gram],
    nacl:[5,tyto.OM.gram],
    yeast_extract:[5,tyto.OM.gram]}
```

I could develop functions to make reactives or fungibles specifying identity and a URL that may need some recommendations or best practices to be more standard.

Then the function that creates media looks like this:
```
def media(identity: str, recipe = None, **kwargs)-> sbol3.Component:
    """
    Creates a media Component of type functional entity.

    :param identity: The identity of the Component.
    :param **kwargs: Keyword arguments of any other Component attribute.
    :return: A Component object.
    """
    media =  sbol3.Component(identity, sbol3.SBO_FUNCTIONAL_ENTITY, **kwargs)
    media.roles.append(tyto.NCIT.Media)
    if recipe:
        for key, value in recipe:
            if isinstance(key, sbol3.Component):
                key = sbol3.SubComponent(key)
            key.measures.append(sbol3.Measure(value[0], value[1]))
            media.features.append(key)
    return media
```

Reply:
This looks like a fine first pass, though I think you might consider externally defined features as well.

Reply:
This function does accept ExternallyDefined features, it just converts Components into SubComponents and if it is ExternallyDefined it just add the measure and append it. I just tested it with an ExternallyDefined object in the recipe. Was that the consideration?

Reply:
Ah, I didn't read carefully enough.  Definitely add an example illustrating this in the regression tests to go with this pull.

Reply:
Sure I can add an example, what and where is the regression tests?  Should I do something like this https://github.com/SynBioDex/SBOL-utilities/blob/develop/test/test_component.py ?

Reply:
Correct. Every significant operation in the utilities should be tested via test functions, which are run automatically in https://github.com/SynBioDex/SBOL-utilities/actions Since you are adding to the component package, adding tests to test_component.py would be good.

Reply:
I implemented some test like the ones that I run on the terminal. I make components using the new functions and add them to the document, then I check if the document is valid or the length of the report == 0.

Reply:
@jakebeal Now the test is working properly, let me know if you have feedback, otherwise we can go ahead with the merge.